Add recover testing to the contact module

GiudGiud opened the pull request:
~~1 test left~~

- add missing prereqs to tests using the same output
- skip a test with no time step
- add an error for an object that clearly does not recover right, skip relevant tests


refs #16857

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

GiudGiud committed:
Recover support for contact module:
- add missing prereqs to tests using the same output
- skip a test with no time step
refs #16857

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2852951/), step Docs: sync website on 43ed283 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30445/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2852966/), step Generate coverage on 43ed283 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>415020</th>
      <th colspan=4>#30445 43ed28</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.46%</td>
      <td>85.46%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112486</td>
      <td>112487</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19136</td>
      <td>19135</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30445/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30445/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30445/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Add error on recover for mortar UO that uses a Dof pointer based map to save old dynamic states
Skip corresponding tests
in recover
Delete file checked in by accident

GiudGiud committed:
Update mortar mesh after displacing the mesh
closes #30451

Regold to displacements changing the dimension of the mesh

moosebuild commented:
Job [Modules recover](https://civet.inl.gov/job/2853327/) on a712fc1 : invalidated by @GiudGiud

did this update or nah

GiudGiud commented:
The regold makes sense for that last commit, but the behavior of the test does not seem right. Can someone check those warnings?

GiudGiud commented:
Ok I looked into it and the additional warnings are simply from the additional call to updateMortarMesh(). So it's ok, every call just gets the warning output.

@lindsayad I won't address recover + distributed contact testing in this PR. And when turning on recover testing for contact, I ll only turn on serial & parallel replicated

GiudGiud commented:
Though if you want to take a look, the 3 failing tests hit the same assert in recover + distributed
```
test:mortar_cartesian_lms.two_block_1st_order_constraint_lm_xy_friction: 
test:mortar_cartesian_lms.two_block_1st_order_constraint_lm_xy_friction: Assertion `interior_parent' failed
test:mortar_cartesian_lms.two_block_1st_order_constraint_lm_xy_friction: No interior parent exists for element 251. There may be a problem with your sideset set-up.
test:mortar_cartesian_lms.two_block_1st_order_constraint_lm_xy_friction: at /Users/giudgl/projects/moose_v8/framework/src/constraints/AutomaticMortarGeneration.C, line 1761
test:mortar_cartesian_lms.two_block_1st_order_constraint_lm_xy_friction: [0] /Users/giudgl/projects/moose_v8/framework/src/constraints/AutomaticMortarGeneration.C, line 1761, compiled May 14 2025 at 17:25:27
```

GiudGiud committed:
Skip tests that cannot recover with distributed because they require a serial mesh
(only skipping in recover + distrib), and recover checkpoint is distributed

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1228,7 +1228,11 @@ FEProblemBase::initialSetup()
+  {
     _displaced_problem->updateMesh();
+    // if displacements were applied to the mesh, the mortar mesh should be updated too
+    updateMortarMesh();
```
Someday we might want to enumerate that only the displaced mortar meshes should be updated but since the most common use cases just have mortar running on the displaced mesh, I don't think adding the enumeration is necessarily worth the added code complexity

lindsayad commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -215,6 +216,10 @@ ComputeDynamicWeightedGapLMMechanicalContact::computeQpIProperties()
 void
 ComputeDynamicWeightedGapLMMechanicalContact::timestepSetup()
 {
+  // These dof maps are not recoverable as they are
```
why not?

loganharbour commented on modules/contact/test/tests/multiple_contact_pairs/tests:
```diff
@@ -85,6 +85,8 @@
     abs_zero = 1.0e-4
     rel_err = 1.0e-4
     allow_warnings = true
+    # Automatic pairing method requires a serial mesh and the checkpoint is distributed
+    mesh_mode = 'replicated'
```
recover isn't managed by capabilities unfortunately

GiudGiud replied:
should I duplicate the tests in the meantime?

lindsayad commented:
Note that we'll be able to add distributed recover testing (generally) after https://github.com/libMesh/libmesh/pull/4172 makes its way into MOOSE

GiudGiud committed:
Address Alex's review:
- add comment on why maps of pointers are not recoverable
- duplicate tests for recover & distributed testing that are incompatible at this time for this object

lindsayad reviewed (commented):
To avoid the sqa error, I think it's appropriate to reference particular functionality like recover/distributed in the requirements
